tests/rfc003_erc20_htlc.rs

```diff
@@ -552,4 +560,80 @@ fn given_short_zero_secret_htlc_should_not_redeem() {
         U256::from(600)
     );
     assert_eq!(client.token_balance_of(token_contract, bob), U256::from(0));
+
+    assert_return_data(&client, transaction_receipt, INVALID_SECRET);
+}
+
+#[test]
+fn given_invalid_secret_htlc_should_revert_tx_with_error() {
+    let docker = Cli::default();
+    let secret_vec = vec![
+        0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8,
+        0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8, 0u8,
+    ];
+
+    let (alice, bob, htlc_address, token_contract, token_amount, client, _handle, _container) =
+        erc20_harness(&docker, Erc20HarnessParams::default());
+
+    // Fund erc20 htlc
+    client.sign_and_send(|nonce, gas_price| UnsignedTransaction {
+        nonce,
+        gas_price,
+        gas_limit: U256::from(100_000),
+        to: Some(token_contract),
+        value: U256::from(0),
+        data: Some(
+            Erc20Htlc::transfer_erc20_tx_payload(
+                TokenQuantity(token_amount.into()),
+                Address(htlc_address.into()),
+            )
+            .into(),
+        ),
+    });
+
+    assert_eq!(
+        client.token_balance_of(token_contract, htlc_address),
+        U256::from(400)
+    );
+    assert_eq!(
+        client.token_balance_of(token_contract, alice),
+        U256::from(600)
+    );
+    assert_eq!(client.token_balance_of(token_contract, bob), U256::from(0));
+
+    let transaction_receipt = client.send_data(htlc_address, Some(Bytes(secret_vec)));
+    log::debug!(
+        "used gas ERC20 invalid secret {:?}",
+        transaction_receipt.gas_used
+    );
+
+    assert_eq!(
+        client.token_balance_of(token_contract, htlc_address),
+        U256::from(400)
+    );
+    assert_eq!(client.token_balance_of(token_contract, bob), U256::from(0));
+    assert_eq!(client.token_balance_of(token_contract, bob), U256::from(0));
+
+    assert_return_data(&client, transaction_receipt, INVALID_SECRET);
+}
+
+fn assert_return_data(
+    client: &ParityClient,
+    transaction_receipt: TransactionReceipt,
+    error_code: &str,
+) {
+    let result = client.get_return_data(transaction_receipt);
+    let return_data = result.err().unwrap();
+    let json = json!(format!("{}{}", "Reverted ", error_code));
+    match return_data {
+        Rpc(e) => {
+            asserting(&"contains VM message")
+                .that(&e.message)
+                .contains("VM execution error.");
+            asserting(&"contains revert reason")
+                .that(&e.data.unwrap())
+                .is_equal_to(json);
+        }
+        _ => assert_that(&true).is_false(),
```

Comment:
I think this can be improved :P

I'll have a go once this PR is merged.

Reply:
Yes please 😬